packages/next/next-server/server/render.tsx

```diff
@@ -258,6 +258,7 @@ function renderDocument(
             locales,
             defaultLocale,
             head: React.Children.toArray(docProps.head || [])
+              .filter((elem) => typeof elem?.type === 'string')
```

Comment:
As far as supporting custom components in `<Head>`, we don't need this. Instead, we can remove this `head` property entirely, and replace the SSR'd `<head>` with something like `<head dangerouslySetInnerHTML={{__html: ReactDOM.renderToStaticString(/* ... */) }} />`

Reply:
any security concerns? I can't think of any.

Reply:
I don’t think there’s any *new* concerns. The immediately children would still be controlled by Next, with the exception of the content where `<Head/>` is injected, which is the case today.

Reply:
Hmm... @devknoll I'm going to need my hand held a bit more because I'm not entirely sure where this `<head dangerouslySetInnerHTML />` needs to go for the "SSR'd `<head>`". So far I've considered multiple places, which all seem to need significant refactoring so I don't want to waste time going down the wrong path:

- [_document?](https://github.com/vercel/next.js/blob/canary/packages/next/pages/_document.tsx?L224#L423-L424)
- [next-server/head?](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/lib/head.tsx#L162)

or somewhere within this `render.tsx` file?

Reply:
<img width="367" alt="Screen Shot 2020-11-11 at 9 18 39 AM" src="https://user-images.githubusercontent.com/6808172/98829507-e5437200-23fe-11eb-8c44-ef4a01511c68.png">

Reply:
![SSR](https://user-images.githubusercontent.com/6808172/98831582-5e43c900-2401-11eb-8c48-ab8845cda023.png)

Reply:
The [_document](https://github.com/vercel/next.js/blob/canary/packages/next/pages/_document.tsx?L224#L423-L424) render method is what I'd expect. This `Head` (and indeed, the entire `_document`) component is only run on the server.

Instead of rendering in the current tree, you'd want to create a new one. So instead of:

```jsx
    return (
      <head {...this.props}>
        {/* snipped children */}
      </head>
    )
```

You'd want to start with something like:

```jsx
    // Note: I'm only using `React.Fragment` here because GitHub's lexer doesn't
    // understand the <></> syntax.
    const html =
      ReactDOMServer.renderToString(<React.Fragment>{/* snipped children */}</React.Fragment>)
    return (
      <head {...this.props} dangerouslySetInnerHTML=(__html: html} />
    )
```

> so I'm wondering how to handle both cases using `<head dangerouslySetInnerHTML />`

I don't think we can handle both cases the same way. Instead, we'd handle it for SSR in `_document` and `head-manager` for CSR. The input to `updateHead` is a React tree that we can subsequently render on the client (using `ReactDOMServer.renderToString` or `ReactDOM.render`) to get the elements we need to patch `head` to the correct state.

Definitely feel free to ask more questions and/or push WIP diffs and I'd be happy to help :smile:

Reply:
@devknoll Thoughts on `renderToStaticMarkup` in favor of `renderToString` "as stripping away the extra attributes can save lots of bytes." Or do we need to leverage React's internal id atts?

Reply:
Nope, `renderToStaticMarkup` should be fine.